flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+                                                    streamEdge.getSourceId(), forwardGroup));
+                });
+
+        topologicallySortedChainedStreamNodesMap.forEach(this::setParallelismBasedOnForwardGroup);
+    }
+
+    private void setParallelismBasedOnForwardGroup(
+            StreamNode startNode, List<StreamNode> chainedStreamNodes) {
+        StreamNodeForwardGroup streamNodeForwardGroup =
+                startAndEndNodeIdToForwardGroupMap.get(startNode.getId());
+        // set parallelism for vertices in forward group.
+        if (streamNodeForwardGroup != null && streamNodeForwardGroup.isParallelismDecided()) {
+            chainedStreamNodes.forEach(
+                    streamNode ->
+                            streamNode.setParallelism(
+                                    streamNodeForwardGroup.getParallelism(), true));
+        }
+        if (streamNodeForwardGroup != null && streamNodeForwardGroup.isMaxParallelismDecided()) {
+            chainedStreamNodes.forEach(
+                    streamNode ->
+                            streamNode.setMaxParallelism(
+                                    streamNodeForwardGroup.getMaxParallelism()));
+        }
+    }
+
+    private void generateHashesByStreamNodeId(Integer streamNodeId) {
+        // Generate deterministic hashes for the nodes in order to identify them across
+        // submission if they didn't change.
+        if (hashes.containsKey(streamNodeId)) {
+            return;
+        }
+        for (int i = 0; i < legacyStreamGraphHasher.size(); ++i) {
+            legacyStreamGraphHasher
+                    .get(i)
+                    .generateHashesByStreamNodeId(streamNodeId, streamGraph, legacyHashes.get(i));
+        }
+        Preconditions.checkState(
+                defaultStreamGraphHasher.generateHashesByStreamNodeId(
+                        streamNodeId, streamGraph, hashes),
+                "Failed to generate hash for streamNode with ID '%s'",
+                streamNodeId);
+    }
+
+    /**
+     * Determines whether the chain info is ready to create job vertex based on the following rules:
+     *
+     * <ul>
+     *   <li>The hashes of all upstream nodes of the start node have been generated.
+     *   <li>The vertices of all upstream non-source-chained nodes have been created and executed
+     *       finished.
+     * </ul>
+     *
+     * @param chainInfo the chain info.
+     * @return true if ready, false otherwise.
+     */
+    private boolean isReadyToCreateJobVertex(OperatorChainInfo chainInfo) {
+        Integer startNodeId = chainInfo.getStartNodeId();
+        if (frozenNodeToStartNodeMap.containsKey(startNodeId)) {
+            return false;
+        }
+        StreamNode startNode = streamGraph.getStreamNode(startNodeId);
+        for (StreamEdge inEdges : startNode.getInEdges()) {
+            Integer sourceNodeId = inEdges.getSourceId();
+            if (!hashes.containsKey(sourceNodeId)) {
```

Comment:
> In which case it can happen that the hash of the `sourceNode` has been created, if that the `sourceNode` has finished or it is a chained source?

Currently, this happens in both cases.

Reply:
> When is the hash of a chained source created?

Currently this only happens while the JobGraph is initializing. In `buildAndGetChainEntryPoints`, if it is a chainable source, a hash will be generated for it immediately.

Reply:
> In which case it can happen that the hash of the `sourceNode` has been created, if that the `sourceNode` has finished or it is a chained source?

It seems the checks are redundant; it only needs to verify whether the upstream hash has been generated.